Error: The terraform-provider-azurerm_v3.58.0_x5 plugin crashed!

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

<!--- Please keep this note for the community --->

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->


### Terraform Version

1.4.6

### AzureRM Provider Version

3.58.0

### Affected Resource(s)/Data Source(s)

azurerm_frontdoor, azurerm_app_configuration

### Terraform Configuration Files

```hcl
app_config.tf:
locals {
    infra_path = "/${local.environment}/frontdoor/${local.name}"
    endpoints_list = [ for i in local.deployment_config : i.frontend_endpoint_name ]
    endpoints_list_string = join(",", local.endpoints_list)
    subscription_id = data.azurerm_subscription.current.id
    subscription_id_string="${local.subscription_id}/resourceGroups/${azurerm_resource_group.frontdoor.name}/providers/Microsoft.Network/frontDoors/${azurerm_frontdoor.main.name}/frontendEndpoints"
}


resource "azurerm_app_configuration" "frontdoor_appconf" {
    name                = "appcs-${local.name}"
    resource_group_name = azurerm_resource_group.frontdoor.name
    location            = azurerm_resource_group.frontdoor.location
    tags = local.tags
  
}

resource "azurerm_role_assignment" "appconf_dataowner" {
    scope                = azurerm_app_configuration.frontdoor_appconf.id
    role_definition_name = "App Configuration Data Owner"
    principal_id         = data.azuread_client_config.current.object_id
}

# Let's upload all Azure FrontDoor endpoint ID to app_configuration attached to Azure FrontDoor's resource group
resource "azurerm_app_configuration_key" "frontdoor_enpoint" {
    for_each = local.deployment_config
        configuration_store_id = azurerm_app_configuration.frontdoor_appconf.id
        key                    = "${local.infra_path}/endpoints/endpoint_id/${each.value.frontend_endpoint_name}"
        label                  = each.value.frontend_endpoint_name
        value                  = "${local.subscription_id_string}/${each.value.frontend_endpoint_name}"

        depends_on = [
            azurerm_role_assignment.appconf_dataowner
        ]
}

# Push list of all endpoints names to app configuration
resource "azurerm_app_configuration_key" "frontdoor_endpoints_list" {
        configuration_store_id = azurerm_app_configuration.frontdoor_appconf.id
        key                    = "${local.infra_path}/endpoints/endpoints_list"
        label                  = "endpoints_list"
        value                  = local.endpoints_list_string
        depends_on = [
            azurerm_role_assignment.appconf_dataowner
        ]
}

main.tf:
locals {
  # General naming convention config
  name = "${var.organization}-${var.project}-${local.environment}"
  location = module.governance.location
  
  # Validation of workspace name
  validation = regex("${var.validation_regex}", terraform.workspace)
  
  # Local variable to access deployment config (per workspace config)
  deployment_config = local.environment != "prod" ? var.frontdoor_deployments[terraform.workspace] : var.prod_frontdoor_deployments[terraform.workspace]
  environment =  element(split("-", terraform.workspace), 1)
  
  # All configuration shared between environments will be there
  fd_env_config = local.environment != "prod" ? var.frontdoor_environment_config[local.environment] : var.prod_frontdoor_environment_config[local.environment]

    # Let's define tags here and use them after
  tags = module.governance.tags
  
  # Free tags will be merged with standard tags defined in governacne module
  free_tags = {
    "createmethod "   = "terraform"
    "sourcerepo"      = "IT.foobar"
    "sourceworkspace" = terraform.workspace
    "Contact"         = data.azuread_client_config.current.client_id
  }
}

# Ressource group for Frontdoor
resource "azurerm_resource_group" "frontdoor" {
    name     = "rg-${local.name}"
    location = local.location
    tags = local.tags
}

resource "azurerm_frontdoor" "main" {
  name                                         = "fd-${local.name}"
  resource_group_name                          = azurerm_resource_group.frontdoor.name
  #enforce_backend_pools_certificate_name_check = false

  # This code block is little tricky - we need to go over map of map haches variables "frontdoor_deployments"
  # Some values that in use are optionals that it why we use "try" blocks.
  dynamic "routing_rule" {
      for_each = [for r in local.deployment_config : {
          frontend_endpoint_name = r.frontend_endpoint_name
          accepted_protocols = try(r.accepted_protocols, [ "Https", "Http" ])
          patterns_to_match = try(r.patterns_to_match, ["/*"])
          
          redirect_type = try(r.redirect_type, "Found") 
          custom_path = try(r.custom_path, "/")  
          redirect_protocol = try(r.redirect_protocol, "HttpsOnly")
          custom_host = try(r.custom_host, "foo.bar.net")
          custom_query_string = try(r.custom_query_string, "")
      }]

      content {
        name = routing_rule.value.frontend_endpoint_name
        frontend_endpoints = [ routing_rule.value.frontend_endpoint_name ] 
        accepted_protocols = routing_rule.value.accepted_protocols
        patterns_to_match  = routing_rule.value.patterns_to_match
       
        redirect_configuration {
            redirect_type =  routing_rule.value.redirect_type
            custom_path =  routing_rule.value.custom_path
            redirect_protocol =  routing_rule.value.redirect_protocol
            custom_host = routing_rule.value.custom_host
            custom_query_string = routing_rule.value.custom_query_string
        }
    }
  }

  backend_pool_load_balancing {
    name = "fd-lb-${local.name}"
  }

  backend_pool_health_probe {
    name = "fd-hp-${local.name}"
    protocol = "Https"
    probe_method = "HEAD"
    interval_in_seconds = 30
  }
 
  # Backend pool must be defined for the main endpoint
  backend_pool {
    name = "${terraform.workspace}-foo-bar"
    backend {
      host_header = local.fd_env_config.bp_host_header
      address     = local.fd_env_config.bp_address
      http_port   = local.fd_env_config.bp_http_port
      https_port  = local.fd_env_config.bp_https_port
    }


    load_balancing_name = "fd-lb-${local.name}"
    health_probe_name   = "fd-hp-${local.name}"
  }
  
   frontend_endpoint {
    name      = "fd-${local.name}"
    host_name = "fd-${local.name}.azurefd.net"
  }

  # Dynamic front-end based on variable "frontdoor_deployments"
  dynamic "frontend_endpoint" {
    for_each = local.deployment_config
     content {
        name      = frontend_endpoint.value.frontend_endpoint_name
        host_name = "${frontend_endpoint.value.frontend_endpoint_name}.foobar.com"

     }
  }

  tags = local.tags
}


# All afdverify DNS records
resource "azurerm_dns_cname_record" "afdverify_main" {
        name                = "afdverify.fd-${local.name}"
        zone_name           = data.azurerm_dns_zone.prod.name
        resource_group_name = local.fd_env_config.dns_rg_name
        ttl                 = 3600 # I am hardcoding this for now
        record              = "afdverify.fd-${local.name}.azurefd.net"
        provider            = azurerm.prod
}

resource "azurerm_dns_cname_record" "afdverify_redirect" {
    for_each = local.deployment_config
        name                = "afdverify.${each.value.frontend_endpoint_name}"
        zone_name           = data.azurerm_dns_zone.prod.name
        resource_group_name = local.fd_env_config.dns_rg_name
        ttl                 = try(each.value.ttl_adverify, 3600)
        record              = azurerm_dns_cname_record.afdverify_main.record
        provider            = azurerm.prod
}

# DNS records for endpoint redirctions
resource "azurerm_dns_cname_record" "redirect" {
    for_each = local.deployment_config
        name                = each.value.frontend_endpoint_name
        zone_name           = data.azurerm_dns_zone.prod.name
        resource_group_name = local.fd_env_config.dns_rg_name
        ttl                 = try(each.value.ttl, 3600)
        record              = "fd-${local.name}.azurefd.net"
        provider            = azurerm.prod
}

input.tf:
variable "frontdoor_deployments" {
    default = {
      "frontdoor-sbx" = {
        foo_bar-redirect = {
            frontend_endpoint_name = "foobar"
            custom_path = "/"
            custom_host = "foobat.net"
            custom_query_string = "selectedProjectType=software"
          }
       }
    }
  }

provider.tf:
terraform {
  backend "azurerm" {
    snapshot = false
    
    # This config will use managed identity to auth to storage account and store tfstate
    # Make sure "Allow storage account key access" is set to "Disabled"
    # Also, this idenity should have "Contributor" RBAC role on storage account
    use_azuread_auth = true
    # This can also be sourced from the ARM_USE_AZUREAD environment variable
    
    # Uncomment to use locally in sandbox
     tenant_id            = "vvvvvvvvvvvvvvvvvvvvvvvvvvvv"
     subscription_id      = "vvvvvvvvvvvvvvvvvvvvvvvvvv"
     resource_group_name  = "vvvvvvvvvvvvvvvvvvvvvvv"
     storage_account_name = "vvvvvvvvvvvvv"
     container_name       = "vvvvvvvvvvvvvvvvvvvv"
     key                  = "terraform.vvvvvvv.tfstate"
  }

  required_providers {
    azuread = {
      source  = "hashicorp/azuread"
      version = "2.39.0"
    }

    azurerm = {
      source  = "hashicorp/azurerm"
      version = "3.58.0"
    }

  }
  required_version = ">= 0.13"
}

provider "azuread" {
  # Configuration options
}

provider "azurerm" {
  # Configuration options
  features {}
}


data.tf:
data "azuread_client_config" "current" {}
data "azurerm_subscription" "current" {}


data "azurerm_dns_zone" "prod" {
  provider            = azurerm.prod
  name                = "foobar.com"
  resource_group_name = "rg-foobar"
}
```


### Debug Output/Panic Output

```shell
Error: Plugin did not respond
│ 
│   with azurerm_app_configuration.frontdoor_appconf,
│   on app_configuration.tf line 17, in resource "azurerm_app_configuration" "frontdoor_appconf":
│   17: resource "azurerm_app_configuration" "frontdoor_appconf" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵
╷
│ Error: Request cancelled
│ 
│   with azurerm_frontdoor.main,
│   on main.tf line 35, in resource "azurerm_frontdoor" "main":
│   35: resource "azurerm_frontdoor" "main" {
│ 
│ The plugin.(*GRPCProvider).ApplyResourceChange request was cancelled.
╵

Stack trace from the terraform-provider-azurerm_v3.58.0_x5 plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x53b31fe]

goroutine 229 [running]:
github.com/hashicorp/terraform-provider-azurerm/internal/services/frontdoor.resourceFrontDoorUpdate(0x0?, {0x6329e80?, 0xc001a14400})
        github.com/hashicorp/terraform-provider-azurerm/internal/services/frontdoor/frontdoor_resource.go:156 +0x49e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0x77f7ec0?, {0x77f7ec0?, 0xc00364c6c0?}, 0xd?, {0x6329e80?, 0xc001a14400?})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/resource.go:729 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc000feddc0, {0x77f7ec0, 0xc00364c6c0}, 0xc003610f70, 0xc00452c380, {0x6329e80, 0xc001a14400})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/resource.go:847 +0x83a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc001292e10, {0x77f7ec0?, 0xc00364c5a0?}, 0xc00324ec80)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/grpc_provider.go:1021 +0xe8d
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc0012c1720, {0x77f7ec0?, 0xc003609d70?}, 0xc002f8f500)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/tf5server/server.go:818 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x6afeba0?, 0xc0012c1720}, {0x77f7ec0, 0xc003609d70}, 0xc002f8f490, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0001de1e0, {0x7807fc0, 0xc0001721a0}, 0xc00360ca20, 0xc001b72f60, 0xbdd4f60, 0x0)
        google.golang.org/grpc@v1.51.0/server.go:1340 +0xd23
google.golang.org/grpc.(*Server).handleStream(0xc0001de1e0, {0x7807fc0, 0xc0001721a0}, 0xc00360ca20, 0x0)
        google.golang.org/grpc@v1.51.0/server.go:1713 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        google.golang.org/grpc@v1.51.0/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.51.0/server.go:963 +0x28a
```


### Expected Behaviour

following resources must be deployed:
azurerm_frontdoor
azurerm_app_configuration
azurerm_role_assignment
azurerm_app_configuration_key
azurerm_app_configuration_key


### Actual Behaviour

Crash messages during apply:

 Error: Plugin did not respond
│ 
│   with azurerm_app_configuration.frontdoor_appconf,
│   on app_configuration.tf line 17, in resource "azurerm_app_configuration" "frontdoor_appconf":
│   17: resource "azurerm_app_configuration" "frontdoor_appconf" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵
╷
│ Error: Request cancelled
│ 
│   with azurerm_frontdoor.main,
│   on main.tf line 35, in resource "azurerm_frontdoor" "main":
│   35: resource "azurerm_frontdoor" "main" {
│ 
│ The plugin.(*GRPCProvider).ApplyResourceChange request was cancelled.

### Steps to Reproduce

terraform apply

### Important Factoids

_No response_

### References

_No response_

Thanks @alexanderogorodnikov for reporting this error. I can reproduce this issue by the following steps:

1. Applying the following config:

```hcl
provider "azurerm" {
  features {}
}

locals {
  name = "f21972"
}

resource "azurerm_resource_group" "example" {
  name     = local.name
  location = "West Europe"
}

resource "azurerm_frontdoor" "example" {
  name                = local.name
  resource_group_name = azurerm_resource_group.example.name

  routing_rule {
    name               = "exampleRoutingRule1"
    accepted_protocols = ["Http", "Https"]
    patterns_to_match  = ["/*"]
    frontend_endpoints = ["exampleFrontendEndpoint1"]
    forwarding_configuration {
      forwarding_protocol = "MatchRequest"
      backend_pool_name   = "exampleBackendBing"
    }
  }

  backend_pool_load_balancing {
    name = "exampleLoadBalancingSettings1"
  }

  backend_pool_health_probe {
    name = "exampleHealthProbeSetting1"
  }

  backend_pool {
    name = "exampleBackendBing"
    backend {
      host_header = "www.bing.com"
      address     = "www.bing.com"
      http_port   = 80
      https_port  = 443
    }

    load_balancing_name = "exampleLoadBalancingSettings1"
    health_probe_name   = "exampleHealthProbeSetting1"
  }

  frontend_endpoint {
    name      = "exampleFrontendEndpoint1"
    host_name = "${local.name}.azurefd.net"
  }
}
```

2. Delete the frontdoor resource on portal.
3. Modify `azurerm_frontdoor.example`, like set `http_port` to `81`.
4. Execute `terraform apply -auto-approve -refresh=false`

It ended with the following error message:

```text
Stack trace from the terraform-provider-azurerm_v3.58.0_x5.exe plugin:

panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x0 pc=0x509c0fe]

goroutine 98 [running]:
github.com/hashicorp/terraform-provider-azurerm/internal/services/frontdoor.resourceFrontDoorUpdate(0x0?, {0x6014440?, 0xc00060
1800})
        github.com/hashicorp/terraform-provider-azurerm/internal/services/frontdoor/frontdoor_resource.go:156 +0x49e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0x74e8da0?, {0x74e8da0?, 0xc001990a20?}, 0xd?, {0
x6014440?, 0xc000601800?})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/resource.go:729 +0x178
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc00117a1c0, {0x74e8da0, 0xc001990a20}, 0xc001dfa
340, 0xc000119b80, {0x6014440, 0xc000601800})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/resource.go:847 +0x83a
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000fa4f48, {0x74e8da0?,
 0xc001990900?}, 0xc001068d20)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.26.1/helper/schema/grpc_provider.go:1021 +0xe8d
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc001037180, {0x74e8da0?, 0xc001953
530?}, 0xc00205a5b0)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/tf5server/server.go:818 +0x574
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x67ea560?, 0xc001
037180}, {0x74e8da0, 0xc001953530}, 0xc00205a4d0, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00023e000, {0x74f8ea0, 0xc001ac2680}, 0xc00295a000, 0xc001aa51a0, 0xbad0ee0
, 0x0)
        google.golang.org/grpc@v1.51.0/server.go:1340 +0xd23
google.golang.org/grpc.(*Server).handleStream(0xc00023e000, {0x74f8ea0, 0xc001ac2680}, 0xc00295a000, 0x0)
        google.golang.org/grpc@v1.51.0/server.go:1713 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        google.golang.org/grpc@v1.51.0/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.51.0/server.go:963 +0x28a

Error: The terraform-provider-azurerm_v3.58.0_x5.exe plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

I'll submit a pr to fix this issue.

--- a/internal/services/compute/linux_virtual_machine_resource.go
+++ b/internal/services/compute/linux_virtual_machine_resource.go
@@ -956,10 +956,6 @@ func resourceLinuxVirtualMachineUpdate(d *pluginsdk.ResourceData, meta interface
 	log.Printf("[DEBUG] Retrieving Linux Virtual Machine %q (Resource Group %q)..", id.Name, id.ResourceGroup)
 	existing, err := client.Get(ctx, id.ResourceGroup, id.Name, compute.InstanceViewTypesUserData)
 	if err != nil {
-		if utils.ResponseWasNotFound(existing.Response) {
-			return nil
-		}
-
 		return fmt.Errorf("retrieving Linux Virtual Machine %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
 	}
 

--- a/internal/services/compute/ssh_public_key_resource.go
+++ b/internal/services/compute/ssh_public_key_resource.go
@@ -150,12 +150,8 @@ func resourceSshPublicKeyUpdate(d *pluginsdk.ResourceData, meta interface{}) err
 		return err
 	}
 
-	existing, err := client.Get(ctx, *id)
+	_, err = client.Get(ctx, *id)
 	if err != nil {
-		if response.WasNotFound(existing.HttpResponse) {
-			return nil
-		}
-
 		return fmt.Errorf("retrieving %s: %+v", *id, err)
 	}
 

--- a/internal/services/compute/windows_virtual_machine_resource.go
+++ b/internal/services/compute/windows_virtual_machine_resource.go
@@ -1006,10 +1006,6 @@ func resourceWindowsVirtualMachineUpdate(d *pluginsdk.ResourceData, meta interfa
 	log.Printf("[DEBUG] Retrieving Windows Virtual Machine %q (Resource Group %q)..", id.Name, id.ResourceGroup)
 	existing, err := client.Get(ctx, id.ResourceGroup, id.Name, compute.InstanceViewTypesUserData)
 	if err != nil {
-		if utils.ResponseWasNotFound(existing.Response) {
-			return nil
-		}
-
 		return fmt.Errorf("retrieving Windows Virtual Machine %q (Resource Group %q): %+v", id.Name, id.ResourceGroup, err)
 	}
 

--- a/internal/services/frontdoor/frontdoor_resource.go
+++ b/internal/services/frontdoor/frontdoor_resource.go
@@ -135,8 +135,7 @@ func resourceFrontDoorUpdate(d *pluginsdk.ResourceData, meta interface{}) error 
 	// remove in 3.0
 	// due to a change in the RP, if a Frontdoor exists in a location other than 'Global' it may continue to
 	// exist in that location, if this is a brand new Frontdoor it must be created in the 'Global' location
-	location := "Global"
-	cfgLocation, hasLocation := d.GetOk("location")
+	var location string
 
 	exists, err := client.Get(ctx, id)
 	if err != nil || exists.Model == nil {
@@ -145,6 +144,7 @@ func resourceFrontDoorUpdate(d *pluginsdk.ResourceData, meta interface{}) error 
 		location = azure.NormalizeLocation(*exists.Model.Location)
 	}
 
+	cfgLocation, hasLocation := d.GetOk("location")
 	if hasLocation {
 		if location != azure.NormalizeLocation(cfgLocation) {
 			return fmt.Errorf("the Front Door %q (Resource Group %q) already exists in %q and cannot be moved to the %q location", name, resourceGroup, location, cfgLocation)

--- a/internal/services/frontdoor/frontdoor_resource.go
+++ b/internal/services/frontdoor/frontdoor_resource.go
@@ -140,9 +140,7 @@ func resourceFrontDoorUpdate(d *pluginsdk.ResourceData, meta interface{}) error 
 
 	exists, err := client.Get(ctx, id)
 	if err != nil || exists.Model == nil {
-		if !response.WasNotFound(exists.HttpResponse) {
-			return fmt.Errorf("locating %s: %+v", id, err)
-		}
+		return fmt.Errorf("locating %s: %+v", id, err)
 	} else {
 		location = azure.NormalizeLocation(*exists.Model.Location)
 	}

